File: src/runtime-prototype/src/StuntDouble/Envelope.hs
{-# LANGUAGE GeneralizedNewtypeDeriving #-}
{-# LANGUAGE DeriveGeneric #-}

module StuntDouble.Envelope where

import Data.ByteString (ByteString)
import Data.Aeson (FromJSON, ToJSON, toEncoding, parseJSON)
import GHC.Generics (Generic)
import Control.Concurrent.STM
import Control.Concurrent.Async

import StuntDouble.Message
import StuntDouble.Reference
import StuntDouble.LogicalTime

------------------------------------------------------------------------

newtype CorrelationId = CorrelationId Int
  deriving (Eq, Ord, Show, Read, Num, Enum, Generic)

instance ToJSON CorrelationId
instance FromJSON CorrelationId

getCorrelationId :: CorrelationId -> Int
getCorrelationId (CorrelationId i) = i

data EnvelopeKind = RequestKind | ResponseKind
  deriving (Eq, Show, Read, Generic)

instance ToJSON EnvelopeKind
instance FromJSON EnvelopeKind

type Envelope = Envelope' Message

data Envelope' msg = Envelope
  { envelopeKind          :: EnvelopeKind
  , envelopeSender        :: RemoteRef
  , envelopeMessage       :: msg
  , envelopeReceiver      :: RemoteRef
  , envelopeCorrelationId :: CorrelationId
  -- | NOTE: We don't need to send the `NodeName` part of `LogicalTime`, only
  -- the integer part over the wire...
  , envelopeLogicalTime   :: LogicalTimeInt
  }
  deriving (Eq, Show, Read, Generic)

-- XXX: We probably want to remove this instance and have a separate envelope
-- codec in the future, which should probably come with a companion message
-- codec.
instance FromJSON msg => FromJSON (Envelope' msg)

replyEnvelope :: Envelope' msg -> Message -> Envelope
replyEnvelope e msg
  | envelopeKind e == RequestKind =
    e { envelopeKind = ResponseKind
      , envelopeSender   = envelopeReceiver e
      , envelopeMessage  = msg
      , envelopeReceiver = envelopeSender e
      }
  | otherwise = error "reply: impossilbe: can't reply to a response..."

File: src/executor-event-loop/envelope.go
package executorEL

import (
	"encoding/json"
)

type EnvelopeKind int

const (
	Request = iota
	Response
)

func (t EnvelopeKind) String() string {
	return [...]string{"RequestKind", "ResponseKind"}[t]
}

func (t *EnvelopeKind) FromString(kind string) EnvelopeKind {
	return map[string]EnvelopeKind{
		"RequestKind":  Request,
		"ResponseKind": Response,
	}[kind]
}

func (t EnvelopeKind) MarshalJSON() ([]byte, error) {
	return json.Marshal(t.String())
}

func (t *EnvelopeKind) UnmarshalJSON(b []byte) error {
	var s string
	err := json.Unmarshal(b, &s)
	if err != nil {
		return err
	}
	*t = t.FromString(s)
	return nil
}

type LocalRef struct {
	Index int `json:"index"`
}

type RemoteRef struct {
	Address string `json:"address"`
	Index   int    `json:"index"`
}

func (r RemoteRef) ToLocal() LocalRef {
	return LocalRef{Index: r.Index}
}

type CorrelationId int
type LogicalTime int

func (lt *LogicalTime) Incr() {
	*lt = *lt + 1
}

func (lt *LogicalTime) Merge(olt LogicalTime) {
	lm := LogicalTime(-1)
	if int(*lt) < int(olt) {
		lm = olt
	} else {
		lm = *lt
	}
	*lt = lm + 1
}

// we need to agree how this should work
type Message struct {
	Kind    string          `json:"kind"`
	Message json.RawMessage `json:"message"` // is json value
}

type Envelope struct {
	Kind          EnvelopeKind  `json:"envelopeKind"`
	Sender        RemoteRef     `json:"envelopeSender"`
	Message       Message       `json:"envelopeMessage"`
	Receiver      RemoteRef     `json:"envelopeReceiver"`
	CorrelationId CorrelationId `json:"envelopeCorrelationId"`
	LogicalTime   LogicalTime   `json:"envelopeLogicalTime"`
}
